Update request: ncmpcpp 0.9.2 → 0.10

- Package name: ncmpcpp
- Latest released version: 0.10
<!-- Search your package here: https://search.nixos.org/packages?channel=unstable -->
- Current version on the unstable channel: 0.9.2
- Current version on the stable/release channel: 0.9.2

<!--
Type the name of your package and try to find an open pull request for the package
If you find an open pull request, you can review it!
There's a high chance that you'll have the new version right away while helping the community!
-->
- [X] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

* Tag: https://github.com/ncmpcpp/ncmpcpp/releases/tag/0.10
* Changelog: https://github.com/ncmpcpp/ncmpcpp/blob/master/CHANGELOG.md#ncmpcpp-010-2024-09-03

I tried pointing the source at the github tag, but the build fails with *"failed to produce output path for output 'out'"*, so the package might need some modifications


**Notify maintainers**

- @k0ral
- @lovek323

<!-- If the search.nixos.org result shows no maintainers, tag the person that last updated the package. -->

-----

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

This seems to break on Darwin (or rather, I believe, clang);

       > clang++ -DHAVE_CONFIG_H -I. -I..   -I/nix/store/2n1l91hj91dxxqp64szkaqxdrp7lmkcw-icu4c-74.2-dev/include -I/nix/store/in244r8shj8kldyx6d7jf036q9d0n5vb-libmpdclient-2.22/include -D_DARWIN_C_SOURCE -DNCURSES_WIDECHAR -I/nix/store/iy590iz2lw27b3fg4k8l9m4p0b6aj91a-ncurses-6.4.20221231-dev/include -I/nix/store/g4zbipp4xgqynsss9sb068hm2khp7jij-curl-8.9.1-dev/include -I/nix/store/nh3k3nvcf59z55ibss2yn82bys976577-taglib-1.13.1/include -I/nix/store/nh3k3nvcf59z55ibss2yn82bys976577-taglib-1.13.1/include/taglib  -g -O2 -flto -ftree-vectorize -ffast-math -std=c++20 -Wall -Wextra -Wshadow -Wimplicit-fallthrough -c -o title.o title.cpp
       > In file included from song_list.cpp:21:
       > ./helpers/song_iterator_maker.h:49:14: error: no template named 'result_of' in namespace 'std'; did you mean 'boost::result_of'?
       >                   typename std::result_of<Extractor(typename IteratorT::reference)>::type,
       >                            ^~~~~~~~~~~~~~
       >                            boost::result_of
       > include/boost/utility/result_of.hpp:72:29: note: 'boost::result_of' declared here
       > template<typename F> struct result_of;
       >                             ^
       > In file included from song_list.cpp:21:
       > ./helpers/song_iterator_maker.h:63:14: error: no template named 'result_of' in namespace 'std'; did you mean 'boost::result_of'?
       >                   typename std::result_of<Extractor(typename ConstIteratorT::reference)>::type,
       >                            ^~~~~~~~~~~~~~
       >                            boost::result_of
       > include/boost/utility/result_of.hpp:72:29: note: 'boost::result_of' declared here
       > template<typename F> struct result_of;
       >                             ^

Maybe we can set `-std=c++17` or something to force a version below c++20, or patch the source?

--- a/pkgs/applications/audio/ncmpcpp/default.nix
+++ b/pkgs/applications/audio/ncmpcpp/default.nix
@@ -1,53 +1,88 @@
-{ lib
-, stdenv
-, fetchurl
-, boost
-, libmpdclient
-, ncurses
-, pkg-config
-, readline
-, libiconv
-, icu
-, curl
-, outputsSupport ? true # outputs screen
-, visualizerSupport ? false, fftw # visualizer screen
-, clockSupport ? true # clock screen
-, taglibSupport ? true, taglib # tag editor
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  boost,
+  libmpdclient,
+  ncurses,
+  pkg-config,
+  readline,
+  libiconv,
+  icu,
+  curl,
+  autoconf,
+  automake,
+  libtool,
+  outputsSupport ? true, # outputs screen
+  visualizerSupport ? false,
+  fftw, # visualizer screen
+  clockSupport ? true, # clock screen
+  taglibSupport ? true,
+  taglib, # tag editor
 }:
 
 stdenv.mkDerivation rec {
   pname = "ncmpcpp";
-  version = "0.9.2";
+  version = "0.10";
 
-  src = fetchurl {
-    url = "https://rybczak.net/ncmpcpp/stable/${pname}-${version}.tar.bz2";
-    sha256 = "sha256-+qv2FXyMsbJKBZryduFi+p+aO5zTgQxDuRKIYMk4Ohs=";
+  src = fetchFromGitHub {
+    owner = "ncmpcpp";
+    repo = "ncmpcpp";
+    rev = "refs/tags/${version}";
+    sha256 = "sha256-HRJQ+IOQ8xP1QkPlLI+VtDUWaI2m0Aw0fCDWHhgsOLY=";
   };
 
   enableParallelBuilding = true;
 
   strictDeps = true;
 
-  configureFlags = [ "BOOST_LIB_SUFFIX=" ]
-    ++ lib.optional outputsSupport "--enable-outputs"
-    ++ lib.optional visualizerSupport "--enable-visualizer --with-fftw"
-    ++ lib.optional clockSupport "--enable-clock"
-    ++ lib.optional taglibSupport "--with-taglib";
+  configureFlags = [
+    "BOOST_LIB_SUFFIX="
+    (lib.enableFeature outputsSupport "outputs")
+    (lib.enableFeature visualizerSupport "enable-visualizer")
+    (lib.withFeature visualizerSupport "fftw")
+    (lib.enableFeature clockSupport "clock")
+    (lib.withFeature taglibSupport "taglib")
+  ];
 
-  nativeBuildInputs = [ pkg-config ]
+  nativeBuildInputs = [
+    autoconf
+    automake
+    libtool
+    pkg-config
+  ];
+
+  buildInputs = [
+    boost
+    libmpdclient
+    ncurses
+    readline
+    libiconv
+    icu
+    curl
+  ] ++ lib.optional visualizerSupport fftw
     ++ lib.optional taglibSupport taglib;
 
-  buildInputs = [ boost libmpdclient ncurses readline libiconv icu curl ]
-    ++ lib.optional visualizerSupport fftw
-    ++ lib.optional taglibSupport taglib;
+  preConfigure =
+    ''
+      ./autogen.sh
+    ''
+    + lib.optionalString stdenv.isDarwin ''
+      # std::result_of was removed in c++20 and unusable for clang16
+      substituteInPlace ./configure \
+        --replace-fail "std=c++20" "std=c++17"
+    '';
 
-  meta = with lib; {
+  meta =  {
     description = "Featureful ncurses based MPD client inspired by ncmpc";
-    homepage    = "https://rybczak.net/ncmpcpp/";
-    changelog   = "https://github.com/ncmpcpp/ncmpcpp/blob/${version}/CHANGELOG.md";
-    license     = licenses.gpl2Plus;
-    maintainers = with maintainers; [ koral lovek323 ];
-    platforms   = platforms.all;
+    homepage = "https://rybczak.net/ncmpcpp/";
+    changelog = "https://github.com/ncmpcpp/ncmpcpp/blob/${version}/CHANGELOG.md";
+    license = lib.licenses.gpl2Plus;
+    maintainers = with lib.maintainers; [
+      koral
+      lovek323
+    ];
+    platforms = lib.platforms.all;
     mainProgram = "ncmpcpp";
   };
 }

